### PR TITLE
Telemetry support for streaming events

### PR DESCRIPTION
**Pull request by zbud-msft**
#### Why I did it

Contains functional changes to stream events via new EventClient
Send heartbeat, configurable via gnmi
Collect stats

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

**Review by renukamanavalan — APPROVED** [review]

**Comment by ganglyu** [comment]
need 80% coverage, please add unit test.

File: sonic_data_client/events_client.go [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+        log.V(4).Infof("Failed to set heartbeat val=%d rc=%d", val, rc)
+    }
+}
+
+func NewEventClient(paths []*gnmipb.Path, prefix *gnmipb.Path, logLevel int) (Client, error) {
+    var evtc EventClient
+    evtc.prefix = prefix
+    for _, path := range paths {
+        // Only one path is expected. Take the last if many
+        evtc.path = path
```
[the remark this answers is not in the document]
Comment (ganglyu) [reply sonic_data_client/events_client.go:88]: If you want to use the last index, it's not necessary to iterate.

File: sonic_data_client/events_client.go [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    }
+
+    for _, e := range evtc.path.GetElem() {
+        keys := e.GetKey()
+        for k, v := range keys {
+            if (k == PARAM_HEARTBEAT) {
+                if val, err := strconv.Atoi(v); err == nil {
+                    log.V(7).Infof("evtc.heartbeat_interval is set to %d", val)
+                    set_heartbeat(val)
```
Comment (ganglyu) [inline sonic_data_client/events_client.go:97]: break?

File: sonic_data_client/events_client.go [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+
+        for _, v := range evtc.last_latencies {
+            if v > 0 {
+                total += v
+                cnt += 1
+            }
+        }
+        if (cnt > 0) {
+            evtc.counters[LATENCY] = (uint64) (total/cnt/1000/1000)
+        }
```
Comment (ganglyu) [inline sonic_data_client/events_client.go:146]: when cnt is 0, what's the LATENCY value?

**Review by ganglyu — COMMENTED** [review]
need 80% coverage, please add unit test.

File: sonic_data_client/events_client.go [hunk 4]
```diff
@@ -0,0 +1,368 @@
+package client
+
+/*
+#cgo CFLAGS: -g -Wall -I../../sonic-swss-common/common -Wformat -Werror=format-security -fPIE 
+#cgo LDFLAGS: -L/usr/lib -lpthread -lboost_thread -lboost_system -lzmq -lboost_serialization -luuid -lswsscommon
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include "events_wrap.h"
```
Comment (ganglyu) [inline sonic_data_client/events_client.go:9]: How do you install this header file?

Comment (zbud-msft) [reply sonic_data_client/events_client.go:9]: We added support for libswsscommon both in yml file and we reference this headerfile by going to ../../sonic-swss-common/common

File: sonic_data_client/events_client.go (hunk 2 above)
Comment (renukamanavalan) [reply sonic_data_client/events_client.go:97]: This is parsing params passed in URL. As expected params is only one, it did not matter. Yet it is good to have a break.

File: sonic_data_client/events_client.go (hunk 3 above)
Comment (renukamanavalan) [reply sonic_data_client/events_client.go:146]: cnt can't be 0. Just realized, as I was re-reading the code. Thanks for pointing out, which helped me re-analyze.

**Comment by renukamanavalan** [comment]
> need 80% coverage, please add unit test.

We will do

**Comment by renukamanavalan** [comment]
Contains functional changes to stream events via new EventClient
Send heartbeat, configurable via gnmi
Collect stats

**Pull request merged**